src/metabase/db/custom_migrations.clj

```diff
@@ -1255,13 +1255,38 @@
-                                                              :select    [:c.id]
-                                                              :where     [:and
-                                                                          [:not :c.archived]
-                                                                          [:= :c.query_type "native"]
-                                                                          [:= :f.id nil]]}])]
-    (doseq [card cards]
-      (update-query-fields! card))))
+  (let [field-ids-for-sql  (requiring-resolve 'metabase.native-query-analyzer/field-ids-for-sql)
+        ;; practically #'metabase.models.query-field/field-ids-for-mbql
+        mbql-fields-xf     (comp
+                            (map #(match %
+                                    ["field" (id :guard integer?) opts] [id (:source-field opts)]
+                                    :else                               nil))
+                            cat
+                            (remove nil?))
+        field-ids-for-mbql (fn [query]
+                             {:direct (->> (tree-seq coll? seq query)
+                                           (into #{} mbql-fields-xf)
+                                           not-empty)})
+        cards              (t2/select :report_card :id [:in {:from      [[:report_card :c]]
+                                                             :left-join [[:query_field :f] [:= :f.card_id :c.id]]
+                                                             :select    [:c.id]
+                                                             :where     [:and
+                                                                         [:not :c.archived]
+                                                                         [:= :f.id nil]]}])]
+    (doseq [{card-id :id query :dataset_query :as card} cards]
+      (let [query                             (json/parse-string query true)
+            {:keys [direct indirect] :as res} (if (= (:type query) "native")
+                                                (field-ids-for-sql query)
+                                                (field-ids-for-mbql query))
+            id->record                        (fn [direct? field-id]
+                                                {:card_id          card-id
+                                                 :field_id         field-id
+                                                 :direct_reference direct?})
+            records                           (concat
+                                               (map (partial id->record true) direct)
+                                               (map (partial id->record false) indirect))
+            known                             (set
+                                               (t2/select-fn-set :id :metabase_field :id [:in (map :field_id records)]))
```

Comment:
redundant `set`
```suggestion
            known                             (t2/select-fn-set :id :metabase_field :id [:in (map :field_id records)])
```

Reply:
Not redundant since `select-fn-set` will return nil if there are no columns, and this migration starts breaking tests for at least one other migration. 🤣